### ISSUE TITLE
No way of passing a timeout from twisted.web.xmlrpc.Proxy to reactor.connectTCP()

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4741]
|<img alt="bigjools's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| bigjools reported|
|-|-|
|Trac ID|trac#4741|
|Type|defect|
|Created|2010-11-17 13:53:30Z|
|Branch|https://github.com/twisted/twisted/tree/xmlrpc-timeout-4741|

Similar to http://twistedmatrix.com/trac/ticket/3450 - there's no way for twisted.web.xmlrpc.Proxy to pass a timeout to reactor.connectTCP().

The default timeout of 30 seconds on connectTCP() is too small for my circumstance.

Attachments:

* [issue_4741.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1e9/1e9850f1c9eb2b355b52f2cdd20b1713e05cd6e3/f6d96ea10e6a5f3571a39ca58aec02855a371aed.patch) (3001 bytes) - added by psykidellic on 2011-02-05 00:35:30Z - Proxy takes timeout parameter which can be passed to underlying connectTCP/connectSSL
* [issue_4741.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1e9/1e9850f1c9eb2b355b52f2cdd20b1713e05cd6e3/1f4e5b19e14cabf729888b8643c4d7b5d2582fab.patch) (3218 bytes) - added by psykidellic on 2011-02-05 00:42:55Z - 
* [issue_4741.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1e9/1e9850f1c9eb2b355b52f2cdd20b1713e05cd6e3/8e26ae4799c06411992f4a46cd20a83116f1e0fa.patch) (4096 bytes) - added by psykidellic on 2011-02-15 00:58:34Z - New patch with better tests
* [issue_4741.4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1e9/1e9850f1c9eb2b355b52f2cdd20b1713e05cd6e3/22e61c6c6661cdf9b4076f41e11ac26d289c2875.patch) (4565 bytes) - added by psykidellic on 2011-02-16 17:01:26Z - Patch with better test case and more informative topfile.
* [issue_4741.5.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1e9/1e9850f1c9eb2b355b52f2cdd20b1713e05cd6e3/2fc18264211381995e4ba3f5b54963a21fec08e8.patch) (4563 bytes) - added by psykidellic on 2011-02-16 17:02:32Z - Removed extra whitespace between classes. Now follows the styling guide better.

<details><summary>Searchable metadata</summary>

```
trac-id__4741 4741
type__defect defect
reporter__bigjools bigjools
priority__normal normal
milestone__ 
branch__branches_xmlrpc_timeout_4741 branches/xmlrpc-timeout-4741
branch_author__psykidellic psykidellic
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1290002010000000 1290002010000000
changetime__1298299073000000 1298299073000000
version__None None
owner__exarkun exarkun
cc__psykidellic cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|psykidellic commented|
|-|-|

After discussing with exarkun, a long term solution of using Agent for XMLRPC would be better. This would allow us to extend to more advanced things. Other was the simpler way of taking a timeout value in Proxy.__init__ and pass it to the underlying connectTCP and connectSSL.


Since I had almost finished writing the patch the simpler way, I am submitting it anyway. You can reject/ignore it if you wish :) I am planning to look into the Agent way soon but this could be a simple middlepath.


Anyway:


1. Update Proxy.__init__, definition to have a timeout keyword argument. Defaults to 30.0, used everywhere.


2. Test case that makes sure the value is passed to the underlying connectXXX methods.


3. Updated topfile.


4. Bump year information.

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|psykidellic removed owner|
|-|-|

issue_4741.2 has the full patch. It applies correctly to rev 30631 of trunk.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_6"></a>|@jerub commented|
|-|-|

(In [30754]) Branching to 'xmlrpc-timeout-4741'

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_7"></a>|@jerub commented|
|-|-|

(In [30756]) Add newsfile

Refs: [#4741](https://github.com/twisted/twisted/issues/4741)

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_8"></a>|@jerub commented|
|-|-|

I've created a branch and committed your change to it.

Your patch adds a new kwarg to t.w.xmlrpc.Proxy, but doesn't update the docstring, please fix that.

I've updated the news file, please check that it's accurate.

I really don't like the test - it's a live-fire test against the reactor, depending on the connection timing out doing a connection to localhost before a connection refused occurs, this is really quite nasty. A better way of testing this behavior is needed.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_9"></a>|@jerub commented|
|-|-|

Whoops, forgot to remove review keyword.

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|psykidellic commented|
|-|-|

Hello

Thanks for the review. I was myself little worried about that test case but I was not sure how to do it.

Now find attached a new patch which applies to this new branch and fixes the above issues.


1. The Proxy did indeed had the docstrings. Just in Proxy: like others, and not with Proxy.__init__(...). I also added docstrings for new reactor that I added for better tests.


2. I updated the topfile, with latest details of reactor argument.


3. Updated the test case as discussed with dash in IRC. Added a reactor argument which helps us in passing a FakeReactor instance in the unittests.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra commented|
|-|-|

Replying to [jerub](#note_8):
> I've updated the news file, please check that it's accurate.

The news file currently reads:
```
twisted.web.xmlrpc.Proxy constructor now takes additional 'timeout' and 'reactor' argument. 'timeout' defaulting to 30 seconds.
```

What about:

```
twisted.web.xmlrpc.Proxy constructor now takes additional 'timeout' and 'reactor' arguments. The 'timeout' argument defaults to 30 seconds.
```

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|psykidellic commented|
|-|-|

Hmm, the docs for reactor was added by me. I will update it to have the correct *.*.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [30809]) Apply issue_4741.5.patch 

refs [#4741](https://github.com/twisted/twisted/issues/4741)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

  1. Naming this new parameter `timeout` will be confusing to people.  It should be `connectTimeout` so people don't think it applies to the request itself.
  1. The `_reactor` attribute docs accidentally document the non-existent `reactor` attribute instead.
  1. The indentation of the `timeout` parameter docs is a little funny, as is the indentation around the `connectSSL/TCP` calls.  Should be four spaces or aligned with the opening `(`.
  1. `FakeReactor` can be `twisted.test.proto_helpers.MemoryReactor` instead, which provides a superset of the functionality.

These are all trivial points, I'll fix them and apply.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [30810]) Address review points

  1. Rename the new `timeout` parameter to `connectTimeout`
  1. Document the `_reactor` attribute by the right name
  1. Fix minor whitespace issues
  1. Use `MemoryReactor` instead of a new `FakeReactor`

refs [#4741](https://github.com/twisted/twisted/issues/4741)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

Ah, one more, pointed out by buildbot, `OpenSSL` is an optional dependency.  The HTTPS part of the test needs to be skipped if it is not available.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

(In [30811]) Split timeout test into separate HTTP and HTTPS test methods.  Skip the HTTPS one if SSL is not available.

refs [#4741](https://github.com/twisted/twisted/issues/4741)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

No Windows slaves online at the moment ([build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/xmlrpc-timeout-4741)) so I'm not going to merge this right away.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set status to `closed`|
|-|-|

(In [30821]) Merge xmlrpc-timeout-4741

Author: psykidellic
Reviewer: jerub, exarkun
Fixes: [#4741](https://github.com/twisted/twisted/issues/4741)

Add a `connectTimeout` parameter to `twisted.web.xmlrpc.Proxy` to control the
connection setup timeout it uses when invoking `reactor.connectTCP` or
`reactor.connectSSL`.
